ports/espressif/common-hal/analogbufio/BufferedIn.c

```diff
@@ -0,0 +1,283 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include "common-hal/analogbufio/BufferedIn.h"
+#include "shared-bindings/analogbufio/BufferedIn.h"
+#include "shared-bindings/microcontroller/Pin.h"
+#include "shared/runtime/interrupt_char.h"
+#include "py/runtime.h"
+#include <string.h>
+#include <stdio.h>
+#include "sdkconfig.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "freertos/semphr.h"
+#include "driver/adc.h"
+
+//#define DEBUG_ANALOGBUFIO
+
+#define NUM_SAMPLES_PER_INTERRUPT   256
+#define NUM_ADC_CHANNELS            1
+#define DMA_BUFFER_SIZE	   	        1024
+#define ATTENUATION                 ADC_ATTEN_DB_0
+#define ADC_READ_TIMEOUT_MS         2000
+
+#if defined(CONFIG_IDF_TARGET_ESP32)
+#define ADC_RESULT_BYTE     2
+#define ADC_CONV_LIMIT_EN   1   //For ESP32, this should always be set to 1
+#elif defined(CONFIG_IDF_TARGET_ESP32S2)
+#define ADC_RESULT_BYTE     2
+#define ADC_CONV_LIMIT_EN   0
+#elif defined(CONFIG_IDF_TARGET_ESP32C3) || defined(CONFIG_IDF_TARGET_ESP32H2)
+#define ADC_RESULT_BYTE     4
+#define ADC_CONV_LIMIT_EN   0
+#elif defined(CONFIG_IDF_TARGET_ESP32S3)
+#define ADC_RESULT_BYTE     4
+#define ADC_CONV_LIMIT_EN   0
+#endif
+
+static adc_digi_convert_mode_t convert_mode = ADC_CONV_SINGLE_UNIT_2;
+static adc_digi_output_format_t output_format = ADC_DIGI_OUTPUT_FORMAT_TYPE1;
+static uint8_t adc_channel = 0;
+
+void common_hal_analogbufio_bufferedin_construct(analogbufio_bufferedin_obj_t *self, const mcu_pin_obj_t *pin, uint32_t sample_rate) {
+    uint16_t adc1_chan_mask = 0;
+    uint16_t adc2_chan_mask = 0;
+
+    if( self->pin != NULL) {
+        mp_raise_ValueError(translate("ADC DMA already initialized"));
+    }
+
+    output_format = ADC_DIGI_OUTPUT_FORMAT_TYPE1;
+    if(pin->adc_index == ADC_UNIT_1) {
+        convert_mode = ADC_CONV_SINGLE_UNIT_1;
+    } else {
+        convert_mode = ADC_CONV_SINGLE_UNIT_2;
+    }
+
+    if (pin->adc_index == NO_ADC || pin->adc_channel == NO_ADC_CHANNEL) {
+        raise_ValueError_invalid_pin();
+    }
+
+    adc_channel = pin->adc_channel;
+
+    /*
+     *	Chip version	Conversion Mode		Output Format Type
+     *	ESP32			1					TYPE1
+     *	ESP32S2			1,2,BOTH,ALTER		TYPE1, TYPE2
+     *	ESP32C3			ALTER				TYPE2
+     *	ESP32S3			1,2,BOTH,ALTER		TYPE2
+     *	ESP32H3			1,2,BOTH,ALTER		TYPE2
+     */
+
+#if defined(CONFIG_IDF_TARGET_ESP32)
+    if(pin->adc_index != ADC_UNIT_1) {
+        mp_raise_ValueError(translate("ESP32 only supports ADC1 unit"));
+    }
```

Comment:
We try to reuse error messages when possible. `mp_raise_ValueError_invalid_pin()` will just say "invalid Pin", which might be enough if we document the limitation to ADC1.

See `locales/circuitpython.pot` for a list of existing error messages.

Reply:
Using the` mp_raise_ValueError_invalid_pin()` in commit  https://github.com/adafruit/circuitpython/pull/7602/commits/ead5751803ae2c770ec5923c5c996cb1f33aa930